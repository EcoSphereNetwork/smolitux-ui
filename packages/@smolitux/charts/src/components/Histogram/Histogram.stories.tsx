import React from 'react';
import { Meta, StoryObj } from '@storybook/react';
import { Histogram } from './Histogram';

const meta: Meta<typeof Histogram> = {
  title: 'Charts/Histogram',
  component: Histogram,
  parameters: { layout: 'centered' },
  tags: ['autodocs'],
  argTypes: {
    height: { control: { type: 'number' } },
    width: { control: { type: 'text' } },
    bins: { control: { type: 'number' } },
  },
};
export default meta;

type Story = StoryObj<typeof Histogram>;

const sampleData = [12, 15, 16, 17, 18, 19, 18, 17, 20, 21, 22, 23];

export const Default: Story = {
  args: {
    data: sampleData,
    bins: 5,
    height: 300,
    width: 600,
    title: 'Sample Histogram',
  },
};
