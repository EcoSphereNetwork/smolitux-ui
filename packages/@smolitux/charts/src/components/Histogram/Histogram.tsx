// ðŸ”§ TODO [Codex]: Tests fehlen â€“ prÃ¼fen & umsetzen
import React, { forwardRef, useMemo } from 'react';
import BarChart, { type BarChartProps, type BarChartSeries } from '../BarChart/BarChart';

export interface HistogramProps extends Omit<BarChartProps, 'data'> {
  /** Numeric values to bin */
  data: number[];
  /** Number of bins */
  bins?: number;
}

/**
 * Histogram visualises frequency distribution of numeric data using BarChart.
 */
export const Histogram = forwardRef<SVGSVGElement, HistogramProps>(
  ({ data, bins = 10, ...rest }, ref) => {
    const barSeries = useMemo<BarChartSeries>(() => {
      if (!data || data.length === 0) {
        return { id: 'hist', name: 'Histogram', data: [] };
      }
      const min = Math.min(...data);
      const max = Math.max(...data);
      const step = (max - min) / bins;
      const counts = new Array(bins).fill(0);
      data.forEach((value) => {
        const index = Math.min(Math.floor((value - min) / step), bins - 1);
        counts[index]++;
      });
      const barData = counts.map((count, i) => {
        const start = min + step * i;
        const end = start + step;
        const label = `${start.toFixed(1)}-${end.toFixed(1)}`;
        return { label, value: count };
      });
      return { id: 'hist', name: 'Histogram', data: barData };
    }, [data, bins]);

    return <BarChart ref={ref} data={barSeries} {...rest} />;
  }
);

Histogram.displayName = 'Histogram';

export default Histogram;
