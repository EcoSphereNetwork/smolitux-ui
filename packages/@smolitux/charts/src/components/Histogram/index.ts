export { default, type HistogramProps } from './Histogram';
