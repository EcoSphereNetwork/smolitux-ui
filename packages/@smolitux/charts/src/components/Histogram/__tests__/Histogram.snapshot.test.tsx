import React from 'react';
import { render } from '@testing-library/react';
import { Histogram } from '../Histogram';

jest.mock('@smolitux/theme', () => ({
  useTheme: () => ({ themeMode: 'light' }),
}));

describe('Histogram Snapshots', () => {
  const values = [1, 2, 2, 3, 4, 5, 5];

  it('renders default chart', () => {
    const { asFragment } = render(<Histogram data={values} bins={4} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
