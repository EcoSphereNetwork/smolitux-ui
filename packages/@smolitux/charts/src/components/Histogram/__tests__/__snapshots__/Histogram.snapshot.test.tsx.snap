// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Histogram Snapshots renders default chart 1`] = `
<DocumentFragment>
  <svg
    class="w-full "
    height="300"
    preserveAspectRatio="xMidYMid meet"
    role="img"
    viewBox="0 0 1000 563"
    width="100%"
  >
    <rect
      fill="#F9FAFB"
      height="230"
      stroke="#E5E7EB"
      stroke-width="1"
      width="100%"
      x="50"
      y="30"
    />
    <g
      class="grid-lines"
    >
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="NaN"
        y1="260"
        y2="260"
      />
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="NaN"
        y1="202.5"
        y2="202.5"
      />
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="NaN"
        y1="145"
        y2="145"
      />
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="NaN"
        y1="87.5"
        y2="87.5"
      />
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="NaN"
        y1="30"
        y2="30"
      />
    </g>
    <g>
      <g>
        <g>
          <rect
            class="animate-rise"
            fill="#3B82F6"
            height="76.66666666666666"
            rx="2"
            stroke="rgba(255,255,255,0.5)"
            stroke-width="1"
            style="transform-origin: bottom; transform: scaleY(0); animation: growUp 0.5s ease-out forwards 0s;"
            width="0"
            x="NaN"
            y="183.33333333333334"
          />
        </g>
      </g>
      <g>
        <g>
          <rect
            class="animate-rise"
            fill="#3B82F6"
            height="153.33333333333331"
            rx="2"
            stroke="rgba(255,255,255,0.5)"
            stroke-width="1"
            style="transform-origin: bottom; transform: scaleY(0); animation: growUp 0.5s ease-out forwards 0s;"
            width="0"
            x="NaN"
            y="106.66666666666669"
          />
        </g>
      </g>
      <g>
        <g>
          <rect
            class="animate-rise"
            fill="#3B82F6"
            height="76.66666666666666"
            rx="2"
            stroke="rgba(255,255,255,0.5)"
            stroke-width="1"
            style="transform-origin: bottom; transform: scaleY(0); animation: growUp 0.5s ease-out forwards 0s;"
            width="0"
            x="NaN"
            y="183.33333333333334"
          />
        </g>
      </g>
      <g>
        <g>
          <rect
            class="animate-rise"
            fill="#3B82F6"
            height="230"
            rx="2"
            stroke="rgba(255,255,255,0.5)"
            stroke-width="1"
            style="transform-origin: bottom; transform: scaleY(0); animation: growUp 0.5s ease-out forwards 0s;"
            width="0"
            x="NaN"
            y="30"
          />
        </g>
      </g>
    </g>
    <g
      class="x-axis"
    >
      <line
        stroke="#9CA3AF"
        stroke-width="1"
        x1="50"
        x2="NaN"
        y1="260"
        y2="260"
      />
      <text
        fill="#4B5563"
        font-size="12"
        text-anchor="middle"
        x="NaN"
        y="280"
      >
        1.0-2.0
      </text>
      <text
        fill="#4B5563"
        font-size="12"
        text-anchor="middle"
        x="NaN"
        y="280"
      >
        2.0-3.0
      </text>
      <text
        fill="#4B5563"
        font-size="12"
        text-anchor="middle"
        x="NaN"
        y="280"
      >
        3.0-4.0
      </text>
      <text
        fill="#4B5563"
        font-size="12"
        text-anchor="middle"
        x="NaN"
        y="280"
      >
        4.0-5.0
      </text>
    </g>
    <g
      class="y-axis"
    >
      <line
        stroke="#9CA3AF"
        stroke-width="1"
        x1="50"
        x2="50"
        y1="30"
        y2="260"
      />
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="45"
          x2="50"
          y1="260"
          y2="260"
        />
        <text
          dominant-baseline="middle"
          fill="#4B5563"
          font-size="12"
          text-anchor="end"
          x="40"
          y="260"
        >
          0
        </text>
      </g>
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="45"
          x2="50"
          y1="202.5"
          y2="202.5"
        />
        <text
          dominant-baseline="middle"
          fill="#4B5563"
          font-size="12"
          text-anchor="end"
          x="40"
          y="202.5"
        >
          0.75
        </text>
      </g>
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="45"
          x2="50"
          y1="145"
          y2="145"
        />
        <text
          dominant-baseline="middle"
          fill="#4B5563"
          font-size="12"
          text-anchor="end"
          x="40"
          y="145"
        >
          1.5
        </text>
      </g>
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="45"
          x2="50"
          y1="87.5"
          y2="87.5"
        />
        <text
          dominant-baseline="middle"
          fill="#4B5563"
          font-size="12"
          text-anchor="end"
          x="40"
          y="87.5"
        >
          2.25
        </text>
      </g>
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="45"
          x2="50"
          y1="30"
          y2="30"
        />
        <text
          dominant-baseline="middle"
          fill="#4B5563"
          font-size="12"
          text-anchor="end"
          x="40"
          y="30"
        >
          3
        </text>
      </g>
    </g>
    <defs>
      <style>
        
            @keyframes growUp {
              to {
                transform: scaleY(1);
              }
            }
            
            @keyframes growRight {
              to {
                transform: scaleX(1);
              }
            }
            
            @keyframes fadeIn {
              to {
                opacity: 1;
              }
            }
          
      </style>
    </defs>
  </svg>
</DocumentFragment>
`;
